Xamarin.Forms.Platform.Android/CollectionView/RecyclerViewScrollListener.cs

```diff
@@ -0,0 +1,116 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Android.Graphics;
+using Android.Support.V7.Widget;
+
+namespace Xamarin.Forms.Platform.Android.CollectionView
+{
+	public class RecyclerViewScrollListener : RecyclerView.OnScrollListener
+	{
+		bool _disposed;
+		int _horizontallOffset, _verticalOffset;
+		ItemsView _itemsView;
+		ItemsViewAdapter _itemsViewAdapter;
+
+		public RecyclerViewScrollListener(ItemsView itemsView, ItemsViewAdapter itemsViewAdapter)
+		{
+			_itemsView = itemsView;
+			_itemsViewAdapter = itemsViewAdapter;
+		}
+
+		public override void OnScrollStateChanged(RecyclerView recyclerView, int newState)
+		{
+			base.OnScrollStateChanged(recyclerView, newState);
+		}
+
+		public override void OnScrolled(RecyclerView recyclerView, int dx, int dy)
+		{
+			base.OnScrolled(recyclerView, dx, dy);
+
+			// TODO: These offsets will be incorrect upon row size or count change.
+			// They are currently provided in place of LayoutManager's default offset calculation
+			// because it does not report accurate values in the presence of uneven rows.
+			// See https://stackoverflow.com/questions/27507715/android-how-to-get-the-current-x-offset-of-recyclerview
+			_horizontallOffset += dx;
+			_verticalOffset += dy;
+
+			var firstVisibleItemIndex = -1;
+			var lastVisibleItemIndex = -1;
+			var centerItemIndex = -1;
+
+			if (recyclerView.GetLayoutManager() is LinearLayoutManager linearLayoutManager)
+			{
+				firstVisibleItemIndex = linearLayoutManager.FindFirstVisibleItemPosition();
+				lastVisibleItemIndex = linearLayoutManager.FindLastVisibleItemPosition();
+				centerItemIndex = CalculateCenterItemIndex(firstVisibleItemIndex, lastVisibleItemIndex, linearLayoutManager);
+			}
+
+			var itemsViewScrolledEventArgs = new ItemsViewScrolledEventArgs
```

Comment:
If I am understanding the code and intent correctly here it is retrieving the index of the item in the center. Given 3 items on screen:
```
[0] [1] [2]
```
The `centerItemIndex` will be 1. Then if the screen continues on showing
```
[2] [3] [4]
```
The `centerItemIndex` will be 3. After performing the calculation it creates the event `ItemsViewScrolledEventArgs`.

It appears that we are sending the event back up the stack with every cycle of the `OnScrolled` leaving the Shared Code to handle when the new item has been scrolled to. With the integration I am attempting this makes it more difficult to handle the `Scrolled` event when the shared code only cares about when the new center item has changed.

@adrianknight89 if everything I stated above is correct, I would propose the following change to this as it is the exact modification I am making to your PR to get through this limitation on my project.

# API Changes
+ Add event to `RecyclerViewScrollListener` - `IndexChanged`
+ Add event to `ItemsView` - `ItemChanged`

# Usage
In the shared code the CarouselView will then add an event listener for `ItemChanged` to perform any logic
```c#
var carouselView = new CarouselView();
carouselView.ItemChanged += (s,e) =>
{
    // business rules for when the item changes
};
```

# Code Snippets
IndexChangedEventArgs
```c#
public class IndexChangedEventArgs : EventArgs
{
    public int Index { get; set; }
    public IndexChangedEventArgs(int index)
    {
        Index = index;
    }
}
```

Change line 45 to this
```c#
var centerItemIndex = CalculateCenterItemIndex(firstVisibleItemIndex, lastVisibleItemIndex, linearLayoutManager);

if (_currentIndex != centerItemIndex)
{
    _currentIndex = centerItemIndex;
    ItemChange?.Invoke(this, new IndexChangedEventArgs(_currentIndex));
}
```
Change line 295 of `ItemsViewRenderer`
```c#
_recycleViewScrollListener = new RecyclerViewScrollListener(ItemsView, ItemsViewAdapter);
_recycleViewScrollListener.ItemChange += (s, e) =>
{
    var index = 0;
    foreach (var item in ItemsView.ItemsSource)
    {
        if (index == e.Index && ItemsView is IItemChanged itemChanged)
        {
            itemChanged.OnItemChanged(new SelectedItemChangedEventArgs(item, index));   
            break;
        }
        index++;
    }
}
AddOnScrollListener(_recycleViewScrollListener);
```

Add the event and create a helper method for invoking the event
```c#
public class ItemsView : View, IItemChanged
{
    // …

    public event EventHandler<SelectedItemChangedEventArgs> ItemChanged;
    void IItemChanged.OnItemChanged(SelectedItemChangedEventArgs eventArgs)
    {
        ItemChanged.Invoke(this, eventArgs);
    }
}
```

@adrianknight89 I understand your PR is trying to accomplish 2 items but 
In my case I added an IItemChanged interface to simplify the invocation in the renderer
```c#
public interface IItemChanged
{
    void OnItemChanged(SelectedItemChangedEventArgs eventArgs);
}
```

## Explanation
The idea here is to only throw an event when the `centerItemIndex` is changed. Once that event is handled in the `ItemsViewRenderer` it can then retrieve the current item. The renderer can then fire an event that includes both 
- IndexChanged
- ItemChanged

This now makes the usage in the shared code very simple

Reply:
@ahoefling I remember thinking about a similar thing at the time of writing the PR, but I couldn't decide if the team would like to see multiple events or a single one to track changes. Remember that `ItemsViewScrolledEventArgs` has multiple properties, and it is very likely for horizontal or vertical offsets to change while visible item indexes remain the same, hence the event firing during every scroll change.

As for your last comment with respect to throwing an event only when `centerItemIndex` changes, I'm not sure if this is the right approach. What happens if the row count changes dynamically such that the center index stays the same while others change? For example, if we have items 0 1 2 & `_currentIndex` is 1, and we remove the last item and scroll the collection, current index should stay the same while last index should be set to 1. What I'm trying to get at is just because `_currentIndex`  remains the same does not mean other indexes have not changed.

Reply:
That said, we could create an `ItemChanged` event and leave offset changes separate and fire `ItemChanged` only when any of the three indexes changes instead of firing it with every offset change. I'd like to hear what the team thinks before moving forward.

Reply:
@adrianknight89 thanks for the detailed clarification, it is helpful for me to understand the intent of the change.

I was only thinking in terms of a CarouselView or horizontal sliding view. In that case you only really need to deal with 1 row of items at a time. On the project that I modified your code I added the event to handle when the center item of a CarouselView has changed.

Perhaps my question/proposal isn't suited for this PR but for a follow-up item that will target the CarouselView specifically (or any 1 row horizontally CollectionView).

I'm curious to hear when the Xamarin Team and others have to say on this as well

Reply:
>Perhaps my question/proposal isn't suited for this PR but for a follow-up item that will target the CarouselView specifically (or any 1 row horizontally CollectionView).

@ahoefling That sounds perfect to me! Please open a new issue. We'll go ahead and merge this one and then the  team and community will be better prepared to discuss your suggestions. Thanks!!!

Reply:
Perfect! I'll reference the PR in the new issue

Reply:
@ahoefling If there is only one item on the screen, then it makes sense to track center item index changes. However, there will be cases when people want to use CarouselView with multiple visible items on the screen because each item is smaller than the width of the screen.

Reply:
@samhouts New issue created: #6756